### ISSUE TITLE
IntelliSense fails to deduce types of CRTP-based expression templates.

### ISSUE THREAD

**Issue by PiliLatiesa** [microsoft/vscode-cpptools#6183]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Linux x86_64
- VS Code Version: 1.49
- C/C++ Extension Version: 1.0.1
- Other extensions you installed (and if the issue persists after disabling them): None
- Does this issue involve using SSH remote to run the extension on a remote machine?: No
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

(Also reported as [DEVCOM-1192188](https://developercommunity.visualstudio.com/content/problem/1192188/intellisense-fails-to-deduce-types-of-crtp-based-e.html))

The following piece of code causes IntelliSense to display red squiggles in the identified line, albeit it is perfectly valid C++. It seems that either the EDG compiler or IntelliSense has trouble in figuring out the type of u + v.

```
template <int dim, int rank, typename, typename>
class BinaryExpr;

template <typename T>
class BaseExpr {};

template <int dim, int rank, typename... TRest, template <int, int, typename...> typename T>
class BaseExpr<T<dim, rank, TRest...>>
{
public:
  template <typename... URest, template <int, int, typename...> typename U>
  BinaryExpr<dim, rank, T<dim, rank, TRest...>, U<dim, rank, URest...>>
  operator +(BaseExpr<U<dim, rank, URest...>> const &rhs) const noexcept
    { return {*this, rhs}; }
};

template <int d, int r, typename T, typename U>
class BinaryExpr : public BaseExpr<BinaryExpr<d, r, T, U>>
{
private:
  BaseExpr<T> const &lhs;
  BaseExpr<U> const &rhs;

public:
  BinaryExpr(BaseExpr<T> const &lhs_, BaseExpr<U> const &rhs_) noexcept :
    lhs(lhs_), rhs(rhs_) {}
};

template <int dim, int rank>
class Field : public BaseExpr<Field<dim, rank>> {};

template <typename T>
void grad(BaseExpr<T> const &) {}

int main()
{
Field<2, 2> u, v;

grad(u + v); // ERROR: no instance of overloaded function matches the specified type
}
```
**Steps to reproduce**

1. Just copy and paste the code in a file. It is a IntelliSense-related problem, independent of c_cpp_properties.json.

**Expected behavior**

IntelliSense to deduce the type of u + v.

<details>
  <summary><strong>Logs</strong></summary>

**Comment by sean-mcmanus** [comment]
Thanks for reporting this. Yes, our parser is shared with VS so we'll track the issue with the VS issue you filed at https://developercommunity.visualstudio.com/content/problem/1192188/intellisense-fails-to-deduce-types-of-crtp-based-e.html .

**Comment by sean-mcmanus** [comment]
@PiliLatiesa The VS team did a big bulk closure of a lot of developer community bugs. We haven't gotten around to refiling them yet. Users impacted could refile if they want. We've started filing new VS bugs internally instead of on the developer community site to avoid the bulk closure problem.

**Comment by sean-mcmanus** [comment]
I've created a new internal bug to track this (1489992).

**Comment by PiliLatiesa** [comment]
The issue has been fixed upstream: [DEVCOM-1192188](https://developercommunity.visualstudio.com/content/problem/1192188/intellisense-fails-to-deduce-types-of-crtp-based-e.html)

Any idea about when the fix will land on the extension?

Thanks

**Comment by bobbrow** [comment]
The fix was just barely committed to VS so it won't be in our 1.19 release because we are nearing release and aren't accepting major IntelliSense updates at this stage in the release cycle (the fix came with EDG's 6.6 release).  Instead, it will make it into our 1.20 release.

**Comment by sean-mcmanus** [comment]
@PiliLatiesa Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.20.0
